ddtrace/internal/datadog/profiling/stack_v2/src/sampler.cpp

```diff
@@ -66,13 +66,26 @@ _stack_v2_atfork_child()
     // The only thing we need to do at fork is to propagate the PID to echion
     // so we don't even reveal this function to the user
     _set_pid(getpid());
+    // We also need to recreate the lock used by echion
+    thread_info_map_lock.~mutex();
+    new (&thread_info_map_lock) std::mutex();
+    {
+        std::lock_guard<std::mutex> lock(thread_info_map_lock);
+        thread_info_map.clear();
```

Comment:
🤔 Is this safe? If the concern is that another thread might have been manipulating `thread_info_map` at the time of fork, couldn't this break as well?

I tried it out with this toy example: https://gist.github.com/nsrip-dd/361372bbd3882532f4fa7167a85e292b. In the toy, the parent creates a map, and then spins off a thread which continually modifies the map. Then the parent forks in a loop, and the child clears the map. With that, the child crashes consistently while clearing the map...

Reply:
To capture what we discussed in person earlier, your idea for how to address this was to have a pre-fork handler in the parent take the needed locks, and then release them post-fork in the parent, so that the child will get the data structures in a clean, protected state.

Looks like we already do this for the `Uploader` object, for example: https://github.com/DataDog/dd-trace-py/blob/d234cdf2b03fc4b05d2c7305b1c689f34fe64774/ddtrace/internal/datadog/profiling/dd_wrapper/src/uploader.cpp#L162-L173.